src/thing_creature.c

```diff
@@ -1241,19 +1255,160 @@ void process_thing_spell_teleport_effects(struct Thing *thing, struct CastedSpel
+                            battleid = 1;
+                        }
+                        for (i = battleid; i <= BATTLES_COUNT; i++)
+                        {
+                            if (i > BATTLES_COUNT)
+                            {
+                                i = 1;
+                                battleid = 1;
+                            }
+                            count++;
+                            struct CreatureBattle* battle = creature_battle_get(i);
+                            if ( (battle->fighters_num != 0) && (battle_with_creature_of_player(thing->owner, i)) )
+                            {
+                                struct Thing* tng = thing_get(battle->first_creatr);
+                                TRACE_THING(tng);
+                                if (creature_can_navigate_to(thing, &tng->mappos, NavRtF_NoOwner))
+                                {
+                                    pos.x.val = tng->mappos.x.val;
+                                    pos.y.val = tng->mappos.y.val;
+                                    battleid = i + 1;
+                                    break;
+                                }
+                            }
+                            if (count >= BATTLES_COUNT)
+                            {
+                                battleid = 1;
+                                break;
+                            }
+                            if (i >= BATTLES_COUNT)
+                            {
+                                i = 0;
+                                battleid = 1;
+                                continue;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        allowed = false;
+                    }
+                    break;
+                }
+                case 17:
+                {
+                    room = room_get(cctrl->last_work_room_id);
+                    break;
+                }
+                case 18:
+                {
+                    struct Coord3d cta_pos;
+                    cta_pos.x.val = subtile_coord_center(dungeon->cta_stl_x);
+                    cta_pos.y.val = subtile_coord_center(dungeon->cta_stl_y);
+                    cta_pos.z.val = subtile_coord(1,0);
+                    if (creature_can_navigate_to(thing, &cta_pos, NavRtF_NoOwner))
+                    {
+                        pos = cta_pos;
+                    }
+                    else
+                    {
+                        allowed = false;
+                    }
+                    break;
+                }
+                case 19:
+                {
+                    desttng = find_hero_door_hero_can_navigate_to(thing);
+                    allowed = (!thing_is_invalid(desttng));
+                    break;
+                }
+                default:
+                {
+                    rkind = zoom_key_room_order[TeleDest-1];
```

Comment:
I thought I already made a comment here, but it is gone. I think you should look into making sure this only happens when teledest-1 can return a room. I've not looked into what values teledest can have, and what happens if it returns -1 or 18 or something.